katdal/applycal.py

```diff
@@ -92,42 +86,50 @@ def complex_interp(x, xi, yi, left=None, right=None):
     return y.astype(yi.dtype)
 
 
-def has_cal_product(cache, attrs, product):
-    """Check if calibration solution `product` is available in `cache`."""
-    key = 'cal_product_' + product
-    try:
-        parts = int(attrs[key + '_parts'])
-    except KeyError:
-        return key in cache
-    else:
-        return any(key + str(part) in cache for part in range(parts))
+def _parse_cal_product(cal_product):
+    """Split `cal_product` into `cal_stream` and `product_type` parts."""
+    fields = cal_product.rsplit('.', 1)
+    product_type = fields.pop()
+    cal_stream = fields.pop() if fields else 'cal'
```

Comment:
Should the default be L1, since the default list is L1.K, L1.B, L1.G?

Reply:
I initially aimed that default at the unit tests to avoid fiddling with them too much. I could remove it here and enforce it on the `visdatav4` level instead.

My thinking is that `visdatav4` is the place to analyse `sdp_archived_streams`, find `sdp.cal` streams and construct "L1" and "L2" combinations, while `applycal` is slightly lower-level and only knows about the stream internals.

Reply:
In that case, maybe the higher level should also supply the default. I'm guessing that at the moment, passing `applycal=['K']` to katdal.open is going to give an error about cal.K not being found?